### ISSUE TITLE
Remember host_db_port

### ISSUE THREAD

**Issue by hchonov** [drud/ddev#3566]
**Is your feature request related to a problem? Please describe.**
Everytime I stop and start a project later it is assigning a different port for accessing the database. It is not possible to assign a specific port as the config might get accidentelly commited, which is not okay in a company where multiple developers work on multiple projects.

**Describe the solution you'd like**
It would be great if DDEV remembers the port assigned when the project has been started for the first time and tries to reuse that port for the same project. It should also reserve it and not assign it to other projects.

**Describe alternatives you've considered**
A ddev config.local.yml which does not get commited? Then we could define there local overrides that each developer likes, e.g. enabling mutagen and specifying static ports for the database.

**Additional context**
No additional context.

**Comment by jonaseberle** [comment]
> A ddev config.local.yml which does not get commited?

Yes, that's how it should work already.

`./.ddev/.gitignore` contains
```
/config.*.y*ml
```

So you can add for example a `./.ddev/config.db-port.yaml` with
```
host_db_port: 33085
```
for your local DB port and it gets gitignored.

**Comment by hchonov** [comment]
I wasn't aware of this feature. Thanks, @jonaseberle . It definetelly works as an alternative solution. But probably it still makes sense for ddev to remember its used ports? :)

**Comment by rfay** [comment]
The reason ddev doesn't try to remember the *localhost* bound ports by default is that different projects can compete with each other for those ports. So ddev might "remember" that it bound port 5000, but a new project might use that. This type of port binding is called "ephemeral ports", where the operating system assigns the port. It can be controlled, but it can lead to quite a big of chaos also. Also, other services can introduce port binding with `docker-compose.*.yaml` that ddev has no control over. And even non-ddev applications may take over ports that ddev would then try to use.

I don't disagree with you that it would be nice if ddev had a master database of exposed ports and always tried to use the same one.

**Comment by jonaseberle** [comment]
The thing is that ddev would need that "try the same port, then if not available ask the user what to do (give the user the opportunity to free up the port and try again or change the port persistently)" logic. That is quite some complexity.

It would be a fine feature but we are fine with how it is now, too. Not every project uses external DB tools and needs a fixed DB port. When the question comes up in our team or in the support channels, the current solution has always been well received (even appraised) and found to be easy enough.

**Comment by hchonov** [comment]
> The thing is that ddev would need that "try the same port, then if not available ask the user what to do (give the user the opportunity to free up the port and try again or change the port persistently)" logic. That is quite some complexity.

I do not think we should make it that complex. If the port is not free then simply assign one dynamically like curretnly is happening. That would not change the user experience or anything else.

**Comment by rfay** [comment]
Most teams that need this just set host_db_port in the checked-in config.yaml, with a different value for each project they're working on.

To understand this better, please explain how you're using the host-bound port, why that's important to you. Note that the `heidisql` and `sequelace` shell commands, among others figure this out automatically, and you can take a look at those and maybe you just want your own custom command for what you're doing?

**Comment by hchonov** [comment]
I use PHPStorm for all my projects and Datagrip inside it. And I usually always check things in the database for all my projects from within Datagrip, but that is not always the case for my co-workers. The same is also in regards to enabling mutagen, which is why we do not commit local stuff to config.yaml.

**Comment by hchonov** [comment]
this is how I configure it and for each project I use the project port.

<img width="803" alt="image" src="https://user-images.githubusercontent.com/8309745/152221080-a285364c-e37a-48a9-9b90-97497685bc61.png">

**Comment by rfay** [comment]
A note that `ddev describe -j | jq -r .raw.dbinfo.published_port` will get the published port.

**Comment by jonaseberle** [comment]
... and the PhpStorm config is in `.idea/dataSources.xml` and looks like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="DataSourceManagerImpl" format="xml" multifile-model="true">
    <data-source source="LOCAL" name="ddev" uuid="9f6345ed-22d8-4732-88a3-a7265d152962">
      <driver-ref>mariadb</driver-ref>
      <synchronize>true</synchronize>
      <jdbc-driver>org.mariadb.jdbc.Driver</jdbc-driver>
      <jdbc-url>jdbc:mariadb://localhost:33085/db</jdbc-url>
    </data-source>
    ...
```
A (custom) `ddev` command could do nice things here ;)

**Comment by hchonov** [comment]
okay, that I something I could do, however by this it will be only me having this and anyone else with a similar requirement will have to it on their own as well, right?

**Comment by rfay** [comment]
The idea of an open-source project is that somebody like you will sort out this and contribute the results back.

**Comment by hchonov** [comment]
I would love to contribute, but I am unsure how? Would such a command be accepted in ddev itself or if not how do I make it available to otherss?

**Comment by rfay** [comment]
I definitely would like that!

**Comment by marcvangend** [comment]
Exactly the same situation here. I would love to have a solution that combines *not* using `host_db_port` (no port collisions, automatically) with easy (automatic or set-and-forget) PhpStorm configuration.

**Comment by rfay** [comment]
I assume you've all used the lovely DDEV integration plugin for PhpStorm, https://plugins.jetbrains.com/plugin/18813-ddev-integration, which makes all this invisible. 

A contribution here is welcome, but it seems like people are using PhpStorm and its database integration just fine these days.
